File: ios/DirectusMobile/WidgetSharedStorage.m
#import <React/RCTBridgeModule.h>

static NSString *const kAppGroup = @"group.com.martijnmichel.directusexpo.widgets";
static NSString *const kConfigListKey = @"directus.widgets.latestItems.v1.configList";
static NSString *const kConfigListFileName = @"configList.json";
static NSString *const kPayloadPrefix = @"directus.widgets.latestItems.v1.payload.";

@interface WidgetSharedStorage : NSObject <RCTBridgeModule>
@end

@implementation WidgetSharedStorage

RCT_EXPORT_MODULE(WidgetSharedStorage)

+ (BOOL)requiresMainQueueSetup {
  return NO;
}

RCT_EXPORT_METHOD(setConfigList:(NSString *)json
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
  NSUserDefaults *defaults = [[NSUserDefaults alloc] initWithSuiteName:kAppGroup];
  if (!defaults) {
    reject(@"APP_GROUP", @"App Group not available", nil);
    return;
  }
  NSString *value = json ?: @"";
  [defaults setObject:value forKey:kConfigListKey];
  [defaults synchronize];
  NSURL *containerURL = [[NSFileManager defaultManager] containerURLForSecurityApplicationGroupIdentifier:kAppGroup];
  if (containerURL) {
    NSURL *fileURL = [containerURL URLByAppendingPathComponent:kConfigListFileName];
    [value writeToURL:fileURL atomically:YES encoding:NSUTF8StringEncoding error:NULL];
  }
  resolve(nil);
}

RCT_EXPORT_METHOD(getConfigListFromAppGroup:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
  NSUserDefaults *defaults = [[NSUserDefaults alloc] initWithSuiteName:kAppGroup];
  if (!defaults) {
    reject(@"APP_GROUP", @"App Group not available", nil);
    return;
  }
  NSString *raw = [defaults stringForKey:kConfigListKey];
  NSUInteger length = raw.length;
  NSInteger count = 0;
  NSMutableArray *ids = [NSMutableArray array];
  if (raw.length > 0 && [raw isKindOfClass:[NSString class]]) {
    NSData *data = [raw dataUsingEncoding:NSUTF8StringEncoding];
    if (data) {
      id json = [NSJSONSerialization JSONObjectWithData:data options:0 error:NULL];
      if ([json isKindOfClass:[NSArray class]]) {
        NSArray *arr = (NSArray *)json;
        count = [arr count];
        for (id item in arr) {
          if ([item isKindOfClass:[NSDictionary class]]) {
            NSString *idVal = [(NSDictionary *)item objectForKey:@"id"];
            if ([idVal isKindOfClass:[NSString class]] && idVal.length > 0) {
              [ids addObject:idVal];
            }
          }
        }
      }
    }
  }
  resolve(@{ @"length": @(length), @"count": @(count), @"ids": ids });
}

RCT_EXPORT_METHOD(setPayload:(NSString *)id
                  json:(NSString *)json
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
  if (id == nil || id.length == 0) {
    reject(@"INVALID_ID", @"config id must be a non-empty string", nil);
    return;
  }
  NSUserDefaults *defaults = [[NSUserDefaults alloc] initWithSuiteName:kAppGroup];
  if (!defaults) {
    reject(@"APP_GROUP", @"App Group not available", nil);
    return;
  }
  NSString *key = [kPayloadPrefix stringByAppendingString:id];
  if (json != nil) {
    [defaults setObject:json forKey:key];
  } else {
    [defaults removeObjectForKey:key];
  }
  [defaults synchronize];
  resolve(nil);
}

@end

File: ios/DirectusMobile/WidgetSharedStorage.swift
import Foundation

private let appGroup = "group.com.martijnmichel.directusexpo.widgets"
private let configListKey = "directus.widgets.latestItems.v1.configList"
private let payloadPrefix = "directus.widgets.latestItems.v1.payload."

@objc(WidgetSharedStorage)
class WidgetSharedStorage: NSObject, RCTBridgeModule {

  static func moduleName() -> String! { "WidgetSharedStorage" }

  static func requiresMainQueueSetup() -> Bool { false }

  @objc
  func setConfigList(_ json: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let defaults = UserDefaults(suiteName: appGroup) else {
      reject("APP_GROUP", "App Group not available", nil)
      return
    }
    defaults.set(json, forKey: configListKey)
    defaults.synchronize()
    resolve(nil)
  }

  @objc
  func setPayload(_ id: String, json: NSString?, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let defaults = UserDefaults(suiteName: appGroup) else {
      reject("APP_GROUP", "App Group not available", nil)
      return
    }
    let key = "\(payloadPrefix)\(id)"
    if let json = json as String? {
      defaults.set(json, forKey: key)
    } else {
      defaults.removeObject(forKey: key)
    }
    defaults.synchronize()
    resolve(nil)
  }

  @objc
  func getConfigListFromAppGroup(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let defaults = UserDefaults(suiteName: appGroup) else {
      reject("APP_GROUP", "App Group not available", nil)
      return
    }

    let raw = defaults.string(forKey: configListKey) ?? ""
    let length = raw.count
    var count = 0
    var ids: [String] = []

    if let data = raw.data(using: .utf8), !raw.isEmpty {
      do {
        if let arr = try JSONSerialization.jsonObject(with: data, options: []) as? [[String: Any]] {
          count = arr.count
          ids = arr.compactMap { ($0["id"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        }
      } catch {
        // Ignore parsing errors; return length with empty ids/count.
      }
    }

    resolve([
      "length": length,
      "count": count,
      "ids": ids,
    ])
  }
}
